common/quotas/global/collection/collection.go

```diff
@@ -163,89 +333,146 @@ func (c *Collection) backgroundUpdateLoop() {
+			// submit local metrics asynchronously, because there's no need to do it synchronously
+			localMetricsDone := make(chan struct{})
+			go func() {
+				defer close(localMetricsDone)
+				c.local.Range(func(k string, v internal.CountedLimiter) bool {
+					globalKey := c.km.LocalToGlobal(k)
+					counts := v.Collect()
+					if counts.Idle > gcAfterIdle || c.shouldDeleteKey(c.keyMode(globalKey), true) {
+						c.logger.Debug(
+							"deleting local ratelimiter",
+							tag.Dynamic("key", globalKey),
+							tag.Dynamic("idle", counts.Idle > gcAfterIdle),
+						)
+						c.local.Delete(k)
+						return true // continue iterating, possibly delete others too
+					}
+
+					c.sendMetrics(globalKey, "local", counts)
+					return true
+				})
+			}()
+
+			capacity := c.global.Len()
 			capacity += capacity / 10 // size may grow while we range, try to avoid reallocating in that case
-			usage := make(map[string]calls, capacity)
-			fallbacks, globals := 0, 0
-			c.limits.Range(func(k string, v *internal.FallbackLimiter) bool {
-				allowed, rejected, usingFallback := v.Collect()
-				if usingFallback {
-					fallbacks++
-				} else {
+			usage := make(map[string]rpc.Calls, capacity)
+			startups, failings, globals := 0, 0, 0
+			c.global.Range(func(k string, v *internal.FallbackLimiter) bool {
+				globalKey := c.km.LocalToGlobal(k)
+				counts, startup, failing := v.Collect()
+				if counts.Idle > gcAfterIdle || c.shouldDeleteKey(c.keyMode(globalKey), false) {
+					c.logger.Debug(
+						"deleting global ratelimiter",
+						tag.Dynamic("key", globalKey),
+						tag.Dynamic("idle", counts.Idle > gcAfterIdle),
+					)
+					c.global.Delete(k)
+					return true // continue iterating, possibly delete others too
+				}
+
+				if startup {
+					startups++
+				}
+				if failing {
+					failings++
+				}
+				if !(startup || failing) {
 					globals++
 				}
-				usage[k] = calls{
-					allowed:  allowed,
-					rejected: rejected,
+				usage[globalKey] = rpc.Calls{
+					Allowed:  counts.Allowed,
+					Rejected: counts.Rejected,
 				}
+				c.sendMetrics(globalKey, "global", counts)
+
 				return true
 			})
+
 			// track how often we're using fallbacks vs non-fallbacks.
-			// this also tells us about how many keys are used per host.
-			// timer-sums unfortunately must be divided by the update frequency to get correct totals.
-			c.scope.RecordTimer(metrics.GlobalRatelimiterFallbackUsageTimer, time.Duration(fallbacks))
-			c.scope.RecordTimer(metrics.GlobalRatelimiterGlobalUsageTimer, time.Duration(globals))
+			// with per-host metrics this can tell us if a host is abnormal
+			// compared to the rest of the cluster, e.g. persistent failing values.
+			c.scope.RecordHistogramValue(metrics.GlobalRatelimiterStartupUsageHistogram, float64(startups))
+			c.scope.RecordHistogramValue(metrics.GlobalRatelimiterFailingUsageHistogram, float64(failings))
+			c.scope.RecordHistogramValue(metrics.GlobalRatelimiterGlobalUsageHistogram, float64(globals))
```

Comment:
I'm not currently sure if we use per-host metrics on... all keys?  some keys?  opt-in?  none?

probably worth finding out, though I don't think it's critical to emit it uniquely per host.  this should only be weird if there's like a split brain or broken host or something, and that seems likely to be identifiable some other way.

Reply:
I think only a few and I'd be hesitant to add a lot more, I think in practice we're pushing our luck with teh load we currently have